torch/csrc/utils/variadic.h

```diff
@@ -108,4 +108,50 @@ struct MakeIndices : MakeIndices<N-1, N-1, Is...> {};
 // 0 through N-1.
 template<size_t... Is>
 struct MakeIndices<0, Is...> { using indices = Indices<Is...>; };
+
+//===----------------------------------------------------------------------===//
+//                                 Utilities
+//===----------------------------------------------------------------------===//
+
+template <bool value, typename T = void>
+using enable_if_t = typename std::enable_if<value, T>::type;
+
+template <bool value, typename T = void>
+using disable_if_t = enable_if_t<!value, T>;
+
+template <typename T>
+using decay_t = typename std::decay<T>::type;
+
+namespace detail {
+template <bool...>
+struct pack;
+} // namespace detail
+
+template <bool... values>
+struct all_of : std::is_same<
+                    detail::pack<values..., true>,
+                    detail::pack<true, values...>> {};
+
+template <bool...>
+struct any_of_t;
+
+template <>
+struct any_of_t<> : std::false_type {};
+
+template <bool head, bool... tail>
+struct any_of_t<head, tail...> {
```

Comment:
`any_of` is basically `std::disjunction` and `all_of` is `std::conjunction`. Maybe makes sense to name them accordingly, and then also take the "official" implementation from [here](http://en.cppreference.com/w/cpp/types/conjunction) to make sure it behaves correctly and we didn't miss any corner cases.

Note: `std::conjunction` and `std::disjunction` have a slightly different interface because they use types instead of bool template arguments, but that actually makes it more versatile. You can still use it with bools together with `std::bool_constant` if you need to.

    template<bool... args> using any_of = std::disjunction<std::bool_constant<args>...>;

Reply:
`std::conjunction` and `std::disjunction` are C++17?

Reply:
Yes, I meant it makes sense to implement our versions of `std::conjunction` and `std::disjunction` instead of implementing a similar-but-different `any_of` and `all_of`. You can take the implementation from cppreference.